src/coding-guidelines/types-and-traits.rst

```diff
@@ -63,3 +64,225 @@ Types and Traits
              current.checked_add(velocity).expect("Position calculation overflowed")
          }
 
+.. guideline:: Do Not Depend on Function Pointer Identity Across Crates
+    :id: gui_QbvIknd9qNF6 
+    :category: required
+    :status: draft
+    :release: unclear-latest
+    :fls: fls_1kg1mknf4yx7
+    :decidability: decidable
+    :scope: system
+    :tags: surprising-behavior
+
+    Do not rely on the equality or stable identity of function pointers originating from different crates or that may be inlined,
+    duplicated, or instantiated differently across compilation units, codegen units, or optimization profiles.
```

Comment:
I'm not sure why you're mentioning optimization profiles here; I don't think it needs to be pointed out that functions may change if you recompile with different compiler settings.

A compilation unit and codegen unit is the same thing for these purposes, and I at least would prefer that people say "codegen unit" because that's what the option is called in the rustc CLI and the Cargo profiles.

Reply:
resolved